### PR TITLE
Make symengine.py classes closer to sympy

### PR DESCRIPTION
**Pull request by mattwala**
A bit of background: I'm working on [sumpy](https://github.com/inducer/sumpy), and we're interested in using symengine as a symbolic backend (currently we use sympy). These changes would help make the symengine transition a bit smoother.

These changes affect:

* Rational
* Derivative
* Subs
* Symbol

cc: @inducer

### CODE REVIEW

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -1349,6 +1374,20 @@ cdef class Subs(Basic):
             m[v_.thisptr] = p_.thisptr
         self.thisptr = symengine.make_rcp_Subs(expr_.thisptr, m)
 
+    @property
+    def expr(self):
+        return self.args[0]
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Following should be faster.
```python
cdef RCP[const symengine.Subs] me = symengine.rcp_static_cast_Subs(self.thisptr)
return c2py(deref(me).get_arg())
```

File: symengine/lib/symengine_wrapper.pyx [hunk 2]
```diff
@@ -1302,6 +1319,14 @@ cdef class Derivative(Basic):
     def is_Derivative(self):
         return True
 
+    @property
+    def expr(self):
+        return self.args[0]
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]:
```python
cdef RCP[const symengine.Derivative] X = symengine.rcp_static_cast_Derivative(self.thisptr)
return c2py(deref(X).get_arg())
```

File: symengine/lib/symengine_wrapper.pyx [hunk 3]
```diff
@@ -387,6 +387,13 @@ def get_dict(*args):
     return D
 
 
+cdef tuple vec_basic_to_tuple(symengine.vec_basic vec):
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: There's a overhead in copying. Passing a reference `symengine.vec_basic&` should fix that.

Comment (mattwala) [inline symengine/lib/symengine_wrapper.pyx]: Done.

File: symengine/lib/symengine_wrapper.pyx (hunk 2 above)
Comment (mattwala) [inline symengine/lib/symengine_wrapper.pyx]: Done (e9aec56c2b8520962994fb52480a73623a7be9fc)

File: symengine/lib/symengine_wrapper.pyx (hunk 1 above)
Comment (mattwala) [inline symengine/lib/symengine_wrapper.pyx]: Done (e9aec56c2b8520962994fb52480a73623a7be9fc)

File: symengine/tests/test_functions.py [hunk 4]
```diff
@@ -86,6 +90,11 @@ def test_Subs():
     assert Subs(Derivative(function_symbol("f", x, y), [x]), [x, y], [_x, x]) \
                 == Subs(Derivative(function_symbol("f", x, y), [x]), [y, x], [x, _x])
 
+    s = Subs(function_symbol("f", _x), [_x], [x])
```
Comment (isuruf) [inline symengine/tests/test_functions.py]: This is not a canonical expression. That's why there's a failure in Debug mode. I created https://github.com/symengine/symengine.py/issues/119 for this.

In the meantime, can you change this to `s = f.diff(x)/2` and change the asserts below accordingly?

Comment (mattwala) [inline symengine/tests/test_functions.py]: Done (81a99aa4ac90d8b6677d808bd6d5e68e4ac1c02f)

File: symengine/lib/symengine_wrapper.pyx [hunk 5]
```diff
@@ -956,6 +965,14 @@ cdef class Rational(Number):
     def is_Rational(self):
         return True
 
+    @property
+    def p(self):
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: To be compatible with SymPy, we need to add `p` and `q` to `Integer` as well.

Comment (mattwala) [inline symengine/lib/symengine_wrapper.pyx]: Done (b38981f85eb26f119f092b518a4ed46fc604305a)

**Pull request merged**